ale/transformation.py

```diff
@@ -0,0 +1,136 @@
+import numpy as np
+
+from ale.rotation import ConstantRotation
+
+class FrameNode():
+    """
+    A single frame in a frame tree. This class is largely adapted from the Node
+    class in the vispy scene graph implementation.
+
+    Attributes
+    __________
+    id : int
+         The NAIF ID code for the frame
+    parent : FrameNode
+             The parent node in the frame tree
+    children : List of FrameNode
+               The children nodes in the frame tree
+    rotation : ConstantRotation or TimeDependentRotation
+               The rotation from this frame to the frame of the parent node
+    """
+
+    def __init__(self, id, parent=None, rotation=None):
+        """
+        Construct a frame node with or without a parent. If a parent is specified
+        then a rotation from this frame to the parent node's frame must be
+        specified and visa-versa.
+
+        Parameters
+        ----------
+        id : int
+             The NAIF ID code for the frame
+        parent : FrameNode
+                 The parent node in the frame chain
+        rotation : ConstantRotation or TimeDependentRotation
+                   The rotation from this frame to the frame of the parent node
+        """
+
+        if (parent is None) != (rotation is None):
+            raise TypeError("parent and rotation must both be entered or both be None.")
+
+        self.children = []
+        self.id = id
+        if parent is not None:
+            self.parent = parent
+        if rotation is not None:
+            self.rotation = rotation
+
+    def __del__(self):
+        """
+        Custom deletor for a FrameNode. The child node is always responsible
+        for updating the parent nodes. So this removes this node from the
+        children of its parent node.
+        """
+        if self.parent is not None:
+            self.parent.children.remove(self)
```

Comment:
I am concerned about the call order here. This will only be called once the reference count goes to zero **and** the garbage collector fires. So, this is a soft guarantee that the child will alert the parent. This might be a case where a context is more appropriate, but I am not sure if a context is workable in this context.